CURATOR-584: Added fault tolerant idempotent Create, SetData and Delete operations

jmslocum16 opened the pull request.

cammckenzie commented:
@jmslocum16 the basic interface looks OK to me, as does the idempotent logic.

Just to confirm my understanding of your solution, your plan is to still have the connection loss exception propagate up to the client code, and they will retry create / update commands with the idempotent flag set, which will succeed if the versions / data match?

jmslocum16 commented:
> @jmslocum16 the basic interface looks OK to me, as does the idempotent logic.
> 
> Just to confirm my understanding of your solution, your plan is to still have the connection loss exception propagate up to the client code, and they will retry create / update commands with the idempotent flag set, which will succeed if the versions / data match?

Hi Cam, thanks for the quick response!
I'm glad I submitted this patch early - having the client retry was actually not the intended behavior. I had assumed the retry loop inside pathInForeground would handle retrying upon connection loss. It sounds like my assumptions were not correct there and i need to dig more into how retrying works, to make sure the client doesn't have to explicitly retry, and write some tests to validate that behavior.

jmslocum16 commented:
Also should I be worried about the CI failing for jdk8 but not jdk11?
It looks unrelated but I'd want to make sure - https://travis-ci.com/github/apache/curator/jobs/459964173

jmslocum16 commented:
@cammckenzie I added additional testing and comments that hopefully help to clarify the intended behavior. It appears that TestCreate uses the RetryOnce retry policy, so CreateBuilder is doing one automatic retry, which is sufficient to demonstrate that it successfully retries on ConnectionLoss.

cammckenzie commented:
@jmslocum16 I probably worded my question poorly. The retry policy will certainly execute a retry on connection loss, but if that retry fails after the configured number of attempts, the error will propagate back to the client which must then do something with it.  I was thinking that perhaps you were suggesting something like the Curator guaranteed deletes, which will retry indefinitely until the zNode in question no longer exists (this approach of retrying in the background indefinitely obviously won't work for create / update cases).

I will try and take a look at your updates today.

jmslocum16 commented:
Ah @cammckenzie I see what you meant! Apologies for misinterpreting. That's correct. If the retry limit is exceeded, the exception will propagate back to the client. I agree that for creates/updates, retrying indefinitely in the background wouldn't make sense, and I think propagating the exception and having the client decide to retry or not is the best approach.

cammckenzie commented:
This looks like a good start to me @jmslocum16. Don't worry about the CI failing for now, some of the unit tests are a bit flaky, will look in more detail before we merge.This looks like a good start to me @jmslocum16. Don't worry about the CI failing for now, some of the unit tests are a bit flaky, will look in more detail before we merge.

jmslocum16 commented:
@cammckenzie Sorry for the delay on this - I was busy with work and the holidays.
I just completed these changes, so they should be good for review whenever is convenient for you. (And apologies for the size of the change, it was a bigger change than anticipated even if ~2/3 of it is test code).

A couple follow-up notes after finishing my implementation:
 - I realized that quietly() for deletes is the same as idempotent(), but I still inherited the Idempotentable interface in DeleteBuilder and reused the quietly() functionality to be consistent with Create and SetData.
 - I didn't implement compareAndDelete or compareAndSetData since those aren't as widely useful as the main Create/SetData/Delete. I could do so in a new PR if desired.
 - I did not add idempotent operations to the x-async framework, since this change was already getting large and i'm not sure what the x-async framework is used for. I could do that in a new PR if desired.
 - I did not change any of the higher-level recipes to use idempotent operations to simplify their implementation. I could do this in a new PR if desired.
 - I did not add idempotent operations to transactions, since I don't think it makes sense to do so.

cammckenzie commented:
Thanks @jmslocum16, will try and take a look in the next few day.

cammckenzie commented:
@jmslocum16, looks good. Some minor comments, but looks pretty close to mergeable to me.

jmslocum16 commented:
> @jmslocum16, looks good. Some minor comments, but looks pretty close to mergeable to me.

Thanks @cammckenzie! You mentioned some minor comments, but I can't find any comments you left. Is there something with the MR UI I'm missing, or are there more comments you're planning on adding?

cammckenzie commented:
@jmslocum16 Forgot to complete my review, so the comments were all in a pending state, they should be there now.

jmslocum16 commented on pom.xml:
```diff
@@ -68,7 +68,7 @@
         <!-- versions -->
         <zookeeper-version>3.6.0</zookeeper-version>
-        <maven-bundle-plugin-version>4.1.0</maven-bundle-plugin-version>
+        <maven-bundle-plugin-version>5.1.1</maven-bundle-plugin-version>
```
Yeah, I changed this because i was getting build errors trying to build curator locally.
This was the recommended fix I found, and it did fix it for me. I am also not a maven expert so if there is a different or better way to fix it I'm happy to change it.

eolivelli replied:
It should be fine to make the upgrade

eolivelli commented on curator-framework/src/main/java/org/apache/curator/framework/imps/CreateBuilderImpl.java:
```diff
@@ -87,6 +93,7 @@ public CreateBuilderImpl(CuratorFrameworkImpl client, CreateMode createMode, Bac
         this.createParentsAsContainers = createParentsAsContainers;
         this.compress = compress;
         this.setDataIfExists = setDataIfExists;
+        this.idempotent = false; // TODO should I add this to async framework as well? Looks like that's the reason this is a public constructor
```
can we please remove TODO and create an issue ? 
or we can leave here the TODO but put the link to the JIRA

jmslocum16 replied:
created https://issues.apache.org/jira/browse/CURATOR-589

eolivelli commented on curator-framework/src/main/java/org/apache/curator/framework/imps/CreateBuilderImpl.java:
```diff
@@ -807,6 +826,53 @@ public void performBackgroundOperation(OperationAndData<PathAndBytes> op) throws
+                }
+                catch ( KeeperException e )
+                {
+                    // ignore
```
log ?

eolivelli commented on curator-framework/src/main/java/org/apache/curator/framework/imps/IdempotentUtils.java:
```diff
@@ -0,0 +1,53 @@
+     */
+    static boolean matches(byte[] expectedData, byte[] actualData)
+    {
+        if ( expectedData.length != actualData.length )
```
can we use java.util.Arrays.equals ?

eolivelli commented on curator-framework/src/main/java/org/apache/curator/framework/imps/SetDataBuilderImpl.java:
```diff
@@ -51,6 +62,8 @@ public SetDataBuilderImpl(CuratorFrameworkImpl client, Backgrounding backgroundi
         this.backgrounding = backgrounding;
         this.version = version;
         this.compress = compress;
+        // TODO jslocum - this is only public for x-async api. Should I add idempotent there too?
```
can we link a JIRA ?

eolivelli commented on curator-framework/src/main/java/org/apache/curator/framework/imps/SetDataBuilderImpl.java:
```diff
@@ -210,6 +230,47 @@ public Stat forPath(String path) throws Exception
+                }
+                catch ( KeeperException e )
+                {
+                    // ignore
```
log ?

eolivelli reviewed (changes requested):
great work
I left a few minor comments

jmslocum16 commented:
@cammckenzie @eolivelli I think I addressed all of your comments.

eolivelli commented:
I will review again on Monday, thanks for the update

cammckenzie commented:
Looks good to me now, not sure why Travis is complaining, I built locally and ran all tests.

jmslocum16 commented:
> Looks good to me now, not sure why Travis is complaining, I built locally and ran all tests.

@cammckenzie I also ran tests locally and it looked good to me, and travis passed on the last version before I addressed the comments. It looked like the failure was something unrelated:

`
[ERROR] Error occurred in starting fork, check output in log
[ERROR] Process Exit Code: 14
[ERROR] Crashed tests:
[ERROR] org.apache.curator.framework.imps.TestReadOnly
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.awaitResultsDone(ForkStarter.java:532)
....
[ERROR] Caused by: org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?`

What is the expected procedure here? Rerun travis or look into the error more?

eolivelli commented:
Rerun travis

cammckenzie commented:
Looks like Travis failed again, I have zero Travis experience unfortunately. Is this a memory issue? Or something else?

[ERROR] Error occurred in starting fork, check output in log
[ERROR] Process Exit Code: 14
[ERROR] Crashed tests:
[ERROR] org.apache.curator.framework.recipes.cache.TestCuratorCacheEdges
[ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: ExecutionException The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was /bin/sh -c cd /home/travis/build/apache/curator/curator-recipes && /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java -jar /home/travis/build/apache/curator/curator-recipes/target/surefire/surefirebooter7288833367410099534.jar /home/travis/build/apache/curator/curator-recipes/target/surefire 2021-02-01T20-48-51_335-jvmRun1 surefire753304824818451703tmp surefire_618215176770182383564tmp
[ERROR] Error occurred in starting fork, check output in log

Randgalt commented:
> Looks like Travis failed again, I have zero Travis experience unfortunately. Is this a memory issue? Or something else?
> [ERROR] Error occurred in starting fork, check output in log

Our tests are extremely flakey. They always have been.

cammckenzie commented:
@Randgalt the problem doesn't seem to be the test though, from the log it looks like the JVM is crashing in some capacity. The tests have been stable for me locally since you put some work into fixing them a while ago.

TisonKun commented on curator-framework/src/main/java/org/apache/curator/framework/imps/SetDataBuilderImpl.java:
```diff
@@ -36,13 +39,22 @@
     private Backgrounding                   backgrounding;
     private int                             version;
     private boolean                         compress;
+    private boolean                         idempotent;
```
ditto default `false` suggest here.

TisonKun commented on curator-framework/src/main/java/org/apache/curator/framework/imps/SetDataBuilderImpl.java:
```diff
@@ -51,6 +63,8 @@ public SetDataBuilderImpl(CuratorFrameworkImpl client, Backgrounding backgroundi
         this.backgrounding = backgrounding;
         this.version = version;
         this.compress = compress;
+        // TODO implement this in CURATOR-589
```
ditto removal suggestion here.

TisonKun commented on curator-framework/src/main/java/org/apache/curator/framework/imps/SetDataBuilderImpl.java:
```diff
@@ -210,6 +231,48 @@ public Stat forPath(String path) throws Exception
+                            rc = KeeperException.Code.BADVERSION.intValue();
+                        }
+                    }
+                    CuratorEvent event = new CuratorEventImpl(client, CuratorEventType.SET_DATA, rc, path, null, ctx, stat, null, null, null, null, null);
```
```suggestion
                    final CuratorEvent event = new CuratorEventImpl(client, CuratorEventType.SET_DATA, rc, path, null, ctx, stat, null, null, null, null, null);
```

TisonKun reviewed (changes requested):
Nice effort :)

Left comments on quality parts and hopefully go through the correctness parts today.

jmslocum16 commented on curator-framework/src/main/java/org/apache/curator/framework/imps/CreateBuilderImpl.java:
```diff
@@ -642,6 +657,10 @@ else if ( (rc == KeeperException.Code.NODEEXISTS.intValue()) && setDataIfExists
                     {
                         backgroundSetData(client, operationAndData, operationAndData.getData().getPath(), backgrounding);
                     }
+                    else if ( (rc == KeeperException.Code.NODEEXISTS.intValue()) && idempotent && !createMode.isSequential() )
```
Actually, after thinking about this some more, and reading the zookeeper documentation, I don't think it's possible to get a NODE_EXISTS exception with sequential creates (which makes sense now that I think about it):
https://zookeeper.apache.org/doc/r3.3.3/api/org/apache/zookeeper/ZooKeeper.html#create(java.lang.String,%20byte[],%20java.util.List,%20org.apache.zookeeper.CreateMode)

So I just removed the check for createMode.isSequential() entirely.

TisonKun replied:
you're right. sequential always succeeds.

cammckenzie commented on curator-framework/src/main/java/org/apache/curator/framework/imps/CreateBuilderImpl.java:
```diff
@@ -569,7 +583,7 @@ public String forPath(final String givenPath, byte[] data) throws Exception
         }
 
-        final String adjustedPath = adjustPath(client.fixForNamespace(givenPath, createMode.isSequential()));
+        final String adjustedPath = adjustPath(client.fixForNamespace(givenPath, createMode.isnSequential()));
```
isnSequential is presumably a typo?

jmslocum16 replied:
oh, yeah :( fixed

cammckenzie commented:
Looks good to me now. Any further comments from anyone else?

TisonKun commented on curator-framework/src/main/java/org/apache/curator/framework/imps/CreateBuilderImpl.java:
```diff
@@ -74,6 +79,7 @@
         createParentsAsContainers = false;
         compress = false;
         setDataIfExists = false;
+        idempotent = false;
```
if we have a default value, then this is redundant.

TisonKun commented on curator-framework/src/main/java/org/apache/curator/framework/imps/CreateBuilderImpl.java:
```diff
@@ -87,6 +93,7 @@ public CreateBuilderImpl(CuratorFrameworkImpl client, CreateMode createMode, Bac
         this.createParentsAsContainers = createParentsAsContainers;
         this.compress = compress;
         this.setDataIfExists = setDataIfExists;
+        this.idempotent = false;
```
ditto

TisonKun commented on curator-framework/src/main/java/org/apache/curator/framework/imps/SetDataBuilderImpl.java:
```diff
@@ -36,13 +39,22 @@
         backgrounding = new Backgrounding();
         version = -1;
         compress = false;
+        idempotent = false;
```
ditto

TisonKun commented on curator-framework/src/main/java/org/apache/curator/framework/imps/SetDataBuilderImpl.java:
```diff
@@ -51,6 +63,7 @@ public SetDataBuilderImpl(CuratorFrameworkImpl client, Backgrounding backgroundi
         this.backgrounding = backgrounding;
         this.version = version;
         this.compress = compress;
+        this.idempotent = false;
```
ditto

TisonKun reviewed (approved):
go through the logic and it looks good to me.

few quality suggestions

jmslocum16 commented:
Whoops I accidentally clicked the "close PR" button.
Thanks a bunch for reviewing @cammckenzie @TisonKun and @eolivelli! 
I'll await any new comments from Enrico, and I addressed your quality comments Tison.

cammckenzie commented:
@TisonKun @eolivelli any further comments. If you're happy, I will merge.
cheers

eolivelli reviewed (approved):
LGTM 

Probably it would be good to cut a release. 
It is a very nice feature and it has been quite some time since our latest release.
I will start an email thread

TisonKun commented:
> @TisonKun @eolivelli any further comments. If you're happy, I will merge.
> cheers

@cammckenzie already given my lgtm.

eolivelli commented:
Merged to master branch !
thank you @jmslocum16

cammckenzie commented:
Thanks for all your work getting this over the line @jmslocum16!

jmslocum16 commented:
Thanks to everyone for your helpful comments!
